crates/install-wheel-rs/src/linker.rs

```diff
@@ -536,3 +547,23 @@ fn hardlink_wheel_files(
 
     Ok(count)
 }
+
+/// Copy from `from` to `to`, ensuring that the parent directory is locked.
+fn synchronized_copy(from: &Path, to: &Path, locks: &Locks) -> std::io::Result<()> {
+    // Ensure we have a lock for the directory.
+    let dir_lock = {
+        let mut locks_guard = locks.0.lock().unwrap();
+        locks_guard
+            .entry(to.parent().unwrap().to_path_buf())
+            .or_insert_with(|| Arc::new(Mutex::new(())))
+            .clone()
+    };
```

Comment:
@konstin - did you have some idea of only using a lock per _top-level_ directory here? Why do that?

Reply:
I think the idea is to hold the lock longer?

> We keep that lock until we have a new directory, so when the next files are `foo/bar/b.py` and `foo/baz/c.py`, we already have the lock and it's just a prefix check.

Reply:
Yeah, it was a micro-optimization